elastic/kibana
# feat(slo): introduce SLO transform installer

## 📝  Summary

Resolves https://github.com/elastic/kibana/issues/139426

💬 Suggest to review commit per commit

This PR changes the index pattern used to match the [convention](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). 
It also introduces the `TransformInstaller` service that creates a SLO specific transform, and starts it. Each SLO type has its own `TransformGenerator`.

💡  The route handler is wired temporarily with all the services. I'd like to extract it within a service so the whole flow is easily tested - but this requires slightly more investigation and can be part of https://github.com/elastic/kibana/issues/139428.

## Manual testing

1. Add the kibana feature flag: `xpack.observability.unsafe.slo.enabled: true` in your kibana.dev.yml
2. Start kibana (after ES): `yarn start`
3. Follow instruction to generate some APM metrics: https://github.com/fkanout/elastic-apm-api-alerts-generator
4. Make sure APM server is running, and Elastic APM integration is configured on Kibana
5. Run the above curl query
6. Check in Kibana the Transforms are running
7. Check in Kibana the SLO have been stored into SavedObjects
8. Check indices `slo-observability.sli-v1-default.2022-09-01` are being write into
9. Query rollup data: `POST slo-observability.sli-*/_search`

### SLO Availability

```
curl --request POST \
  --url http://localhost:5601/rtg/api/observability/slos \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "o11y-app GET /flaky services availability",
	"description": "99% o11y-app GET /flaky service availability",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "development",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /flaky"
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.99
	}
}'
```

### SLO Duration


```
curl --request POST \
  --url http://localhost:5601/rtg/api/observability/slos \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"name": "My SLO Service Latency",
	"description": "My SLO Desc",
	"indicator": {
		"type": "slo.apm.transaction_duration",
		"params": {
			"environment": "development",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /slow",
			"threshold.us": 4500000
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.99
	}
}'
```
